dace/transformation/helpers.py

```diff
@@ -17,6 +18,307 @@
+            for src_node in state.nodes():
+                if not isinstance(src_node, nodes.Tasklet):
+                    continue
+                for dst_node in state.nodes():
+                    if src_node is dst_node:
+                        continue
+                    if not isinstance(dst_node, nodes.Tasklet):
+                        continue
+                    for e in state.edges_between(src_node, dst_node):
+                        if e.data.data and e.data.data in sdfg.arrays:
+                            write_set.add(e.data.data)
+        # Add data from edges
+        for src in states:
+            for dst in states:
+                for edge in sdfg.edges_between(src, dst):
+                    for s in edge.data.free_symbols:
+                        if s in sdfg.arrays:
+                            read_set.add(s)
+
+        # Find NestedSDFG's unique data
+        rw_set = read_set | write_set
+        unique_set = set()
+        for name in rw_set:
+            if not sdfg.arrays[name].transient:
+                continue
+            found = False
+            for state in sdfg.states():
+                if state in states:
+                    continue
+                for node in state.nodes():
+                    if (isinstance(node, nodes.AccessNode) and node.data == name):
+                        found = True
+                        break
+            if not found:
+                unique_set.add(name)
+
+        # Find NestedSDFG's connectors
+        read_set = {n for n in read_set if n not in unique_set or not sdfg.arrays[n].transient}
+        write_set = {n for n in write_set if n not in unique_set or not sdfg.arrays[n].transient}
+
+        # Find defined subgraph symbols
+        defined_symbols = set()
+        strictly_defined_symbols = set()
+        for e in subgraph.edges():
+            defined_symbols.update(set(e.data.assignments.keys()))
+            for k, v in e.data.assignments.items():
+                try:
+                    if k not in sdfg.symbols and k not in {str(a) for a in symbolic.pystr_to_symbolic(v).args}:
+                        strictly_defined_symbols.add(k)
+                except AttributeError:
+                    # `symbolic.pystr_to_symbolic` may return bool, which doesn't have attribute `args`
+                    pass
+
+        new_state = sdfg.add_state('nested_sdfg_parent')
+        nsdfg = SDFG("nested_sdfg", constants=sdfg.constants_prop, parent=new_state)
+        nsdfg.add_node(source_node, is_start_state=True)
+        nsdfg.add_nodes_from([s for s in states if s is not source_node])
+        for s in states:
+            s.parent = nsdfg
+        for e in subgraph.edges():
+            nsdfg.add_edge(e.src, e.dst, e.data)
+
+        for e in sdfg.in_edges(source_node):
+            sdfg.add_edge(e.src, new_state, e.data)
+        for e in sdfg.out_edges(sink_node):
+            sdfg.add_edge(new_state, e.dst, e.data)
```

Comment:
if we make a new sink node, we can aggregate all of the out_edges of each sink node of the original subgraph here

Reply:
This needs to be studied further. The main purpose of the method is to nest scopes, not arbitrary subgraphs. Scopes should always have a unique sink. Of course, the DeadStateElimination pass breaks this, but we cannot solve all those problems together.